File: homework2/Csharp/OVER/OperatorOverloading.cs
using System;

namespace OperatorOverloadingExample
{
    // 간단한 2D 벡터 클래스 정의
    public class Vector2D
    {
        public double X { get; set; }
        public double Y { get; set; }

        // 생성자
        public Vector2D(double x, double y)
        {
            X = x;
            Y = y;
        }

        // + 연산자 오버로딩 (벡터 덧셈)
        public static Vector2D operator +(Vector2D v1, Vector2D v2)
        {
            return new Vector2D(v1.X + v2.X, v1.Y + v2.Y);
        }

        // - 연산자 오버로딩 (벡터 뺄셈)
        public static Vector2D operator -(Vector2D v1, Vector2D v2)
        {
            return new Vector2D(v1.X - v2.X, v1.Y - v2.Y);
        }

        // * 연산자 오버로딩 (스칼라 곱셈)
        public static Vector2D operator *(Vector2D v, double scalar)
        {
            return new Vector2D(v.X * scalar, v.Y * scalar);
        }

        // / 연산자 오버로딩 (스칼라 나눗셈)
        public static Vector2D operator /(Vector2D v, double scalar)
        {
            if (scalar == 0)
                throw new DivideByZeroException("스칼라 값은 0이 될 수 없습니다.");
            return new Vector2D(v.X / scalar, v.Y / scalar);
        }

        // 벡터를 문자열로 변환 (출력용)
        public override string ToString()
        {
            return $"({X}, {Y})";
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            // 두 벡터 생성
            Vector2D v1 = new Vector2D(3, 4);
            Vector2D v2 = new Vector2D(1, 2);

            // 연산자 오버로딩 테스트
            Vector2D addition = v1 + v2; // 벡터 덧셈
            Vector2D subtraction = v1 - v2; // 벡터 뺄셈
            Vector2D multiplication = v1 * 2; // 스칼라 곱셈
            Vector2D division = v1 / 2; // 스칼라 나눗셈

            // 결과 출력
            Console.WriteLine($"v1: {v1}");
            Console.WriteLine($"v2: {v2}");
            Console.WriteLine($"덧셈: {v1} + {v2} = {addition}");
            Console.WriteLine($"뺄셈: {v1} - {v2} = {subtraction}");
            Console.WriteLine($"스칼라 곱셈: {v1} * 2 = {multiplication}");
            Console.WriteLine($"스칼라 나눗셈: {v1} / 2 = {division}");
        }
    }
}

File: homework2/Kotlin/OVER/OperatorOverloading.kt
// 연산자 오버로딩을 활용한 사칙연산 예제
// 이 예제에서는 Fraction(분수) 클래스에 대해 +, -, *, / 연산자를 오버로딩합니다.

data class Fraction(val numerator: Int, val denominator: Int) {
    // 덧셈 연산자 오버로딩
    operator fun plus(other: Fraction): Fraction {
        // 분모를 통일한 후 분자끼리 더함
        val commonDenominator = denominator * other.denominator
        val newNumerator = numerator * other.denominator + other.numerator * denominator
        return Fraction(newNumerator, commonDenominator).simplify()
    }

    // 뺄셈 연산자 오버로딩
    operator fun minus(other: Fraction): Fraction {
        // 분모를 통일한 후 분자끼리 뺌
        val commonDenominator = denominator * other.denominator
        val newNumerator = numerator * other.denominator - other.numerator * denominator
        return Fraction(newNumerator, commonDenominator).simplify()
    }

    // 곱셈 연산자 오버로딩
    operator fun times(other: Fraction): Fraction {
        // 분자끼리 곱하고 분모끼리 곱함
        val newNumerator = numerator * other.numerator
        val newDenominator = denominator * other.denominator
        return Fraction(newNumerator, newDenominator).simplify()
    }

    // 나눗셈 연산자 오버로딩
    operator fun div(other: Fraction): Fraction {
        // 첫 번째 분수에 두 번째 분수의 역수를 곱함
        val newNumerator = numerator * other.denominator
        val newDenominator = denominator * other.numerator
        return Fraction(newNumerator, newDenominator).simplify()
    }

    // 분수를 간단히 만드는 함수
    private fun simplify(): Fraction {
        val gcd = gcd(numerator, denominator)
        return Fraction(numerator / gcd, denominator / gcd)
    }

    // 최대공약수를 구하는 함수
    private fun gcd(a: Int, b: Int): Int {
        return if (b == 0) a else gcd(b, a % b)
    }

    override fun toString(): String {
        return "$numerator/$denominator"
    }
}

fun main() {
    // Fraction 클래스의 연산자 오버로딩 테스트
    val fraction1 = Fraction(1, 2) // 1/2
    val fraction2 = Fraction(1, 3) // 1/3

    println("Testing Fraction Operations:")
    println("Fraction 1: $fraction1")
    println("Fraction 2: $fraction2")
    println()

    val additionResult = fraction1 + fraction2
    println("Addition:")
    println("$fraction1 + $fraction2 = $additionResult")
    println()

    val subtractionResult = fraction1 - fraction2
    println("Subtraction:")
    println("$fraction1 - $fraction2 = $subtractionResult")
    println()

    val multiplicationResult = fraction1 * fraction2
    println("Multiplication:")
    println("$fraction1 * $fraction2 = $multiplicationResult")
    println()

    val divisionResult = fraction1 / fraction2
    println("Division:")
    println("$fraction1 / $fraction2 = $divisionResult")
    println()
}